### PR TITLE
Remove the globals for GuestPathfinding and Improve Pathfinding tests

### PR DESCRIPTION
**Pull request by LordOfLunacy**
This pull request is to remove the globals gPeepPathFindGoalPosition, gPeepPathFindForeignQueues, and gPeepPathFindQueueRideIndex. Additionally, this pull request will be improving the robustness of tests for pathfinding, particularly for the purpose of confirming that changes to OriginalPathfinding to accommodate the removal of these globals, does not in any way break the current pathfinding.

### CODE REVIEW

**Comment by LordOfLunacy** [comment]
Ok, I am comfortable with this code being added now. I have added a test for guests entering ride queues, though I will likely make it my goal to continue adding tests to pathfinding as I do more work with it. 

Additionally, I ended up changing the interface to not have the SetNewDestination() function that I had added in the initial commit to this PR, and instead placed a function that serves the same functionality in the peep struct called SetPathfindGoal(), since this seems reasonable considering that ResetPathfindGoal() is a member of the struct.

File: src/openrct2-win/test.csv [hunk 1]
```diff
@@ -0,0 +1,113 @@
+function_name;calls;min_microseconds;max_microseconds;average_microseconds
```
Comment (duncanspumpkin) [inline src/openrct2-win/test.csv:1]: ? What is this file doing here

Comment (LordOfLunacy) [reply src/openrct2-win/test.csv:1]: Oops

**Comment by LordOfLunacy** [comment]
Just a follow up, but is this PR still under review/need anything done ?

File: src/openrct2/entity/Staff.cpp [hunk 2]
```diff
@@ -711,12 +711,7 @@ Direction Staff::MechanicDirectionPath(uint8_t validDirections, PathElement* pat
             }
         }
 
-        gPeepPathFindGoalPosition.x = location.x;
-        gPeepPathFindGoalPosition.y = location.y;
-        gPeepPathFindGoalPosition.z = location.z;
-
-        gPeepPathFindIgnoreForeignQueues = false;
-        gPeepPathFindQueueRideIndex = RideId::GetNull();
+        this->SetPathfindGoal(TileCoordsXYZ(location.ToCoordsXYZ()));
```
Comment (Broxzier) [inline src/openrct2/entity/Staff.cpp:714]: We don't use `this->` anywhere in our codebase.

File: src/openrct2/peep/GuestPathfinding.cpp [hunk 3]
```diff
@@ -2312,6 +2293,27 @@ void Peep::ResetPathfindGoal()
...
+    TileCoordsXYZ goal = TileCoordsXYZ(coord);
+    if (this->PathfindGoal != goal)
+    {
+        this->PathfindGoal = { goal, 0 };
+
+        // Clear pathfinding history
+        TileCoordsXYZD nullPos;
+        nullPos.SetNull();
+
+        std::fill(std::begin(this->PathfindHistory), std::end(this->PathfindHistory), nullPos);
```
Comment (Broxzier) [inline src/openrct2/peep/GuestPathfinding.cpp:2307]: Remove them here too.

File: src/openrct2/peep/GuestPathfinding.cpp [hunk 4]
```diff
@@ -855,17 +849,14 @@ static void peep_pathfind_heuristic_search(
                 }
                 else
                 { // numEdges == 2
-                    if (tileElement->AsPath()->IsQueue()
-                        && tileElement->AsPath()->GetRideIndex() != gPeepPathFindQueueRideIndex)
+                    bool rideQueueCheck = tileElement->AsPath()->IsQueue() && peep.Is<Guest>()
+                        && tileElement->AsPath()->GetRideIndex() != peep.As<Guest>()->GuestHeadingToRideId
+                        && !tileElement->AsPath()->GetRideIndex().IsNull();
```
Comment (Broxzier) [inline src/openrct2/peep/GuestPathfinding.cpp:854]: `tileElement->AsPath()` is called a lot here. Store it in a variable instead.

File: test/tests/Pathfinding.cpp [hunk 5]
```diff
@@ -97,7 +98,7 @@ class PathfindingTestBase : public testing::Test
         // destination will be picked, to try and get the peep towards the overall pathfinding goal.
         peep->PeepDirection = moveDir;
         auto destination = CoordsDirectionDelta[moveDir] + peep->GetLocation();
-        peep->SetDestination(destination, 2);
+        peep->SetDestination(destination);
```
Comment (Broxzier) [inline test/tests/Pathfinding.cpp:101]: Any reason for removing the tolerance here?

Comment (LordOfLunacy) [reply test/tests/Pathfinding.cpp:101]: Looking at it, I think I might have removed it during a moment of misunderstanding, I'll add it back in my next commit, along with fixing the other things you've pointed out.

File: src/openrct2/peep/GuestPathfinding.cpp [hunk 6]
```diff
@@ -855,17 +851,14 @@ static void peep_pathfind_heuristic_search(
                 }
                 else
                 { // numEdges == 2
-                    if (tileElement->AsPath()->IsQueue()
-                        && tileElement->AsPath()->GetRideIndex() != gPeepPathFindQueueRideIndex)
+
+                    bool rideQueueCheck = path->IsQueue() && peep.Is<Guest>()
+                        && path->GetRideIndex() != peep.As<Guest>()->GuestHeadingToRideId && !path->GetRideIndex().IsNull();
```
Comment (duncanspumpkin) [inline src/openrct2/peep/GuestPathfinding.cpp:856]: This is not nice to read can you break it up into separate statements.

File: src/openrct2/peep/GuestPathfinding.cpp [hunk 7]
```diff
@@ -1338,7 +1328,7 @@ Direction OriginalPathfinding::ChooseDirection(const TileCoordsXYZ& loc, Peep& p
 
     permitted_edges &= 0xF;
     uint8_t edges = permitted_edges;
-    if (isThin && peep.PathfindGoal == goal)
+    if (isThin)
```
Comment (duncanspumpkin) [inline src/openrct2/peep/GuestPathfinding.cpp:1331]: Why has this changed

Comment (LordOfLunacy) [reply src/openrct2/peep/GuestPathfinding.cpp:1331]: The code became redundant with PathfindGoal no longer being updated through a global variable, however thinking about it, adding a getter method, and making PathfindGoal private within the struct (and adding a getter method) is probably a good idea to make sure only the setter is used to change the PathfindGoal and improving the overall codebase quality.

File: src/openrct2/peep/GuestPathfinding.cpp (hunk 6 above)
Comment (LordOfLunacy) [reply src/openrct2/peep/GuestPathfinding.cpp:856]: Sure, I will do that while I am figuring out why changing the tileElement->AsPath() to a variable as Broxier suggested is causing the warning about potentially dereferencing a nullptr to act up in the Linux bullseye and jammy builds.

Comment (duncanspumpkin) [reply src/openrct2/peep/GuestPathfinding.cpp:856]: Just do a nullptr check after the aspath

**Comment by tupaschoal** [comment]
No worries, thanks for replying!

**Comment by janisozaur** [comment]
I resolved the conflicts. Haven't ran the tests locally yet.

**Comment by LordOfLunacy** [comment]
I apologize, I never finished this pull request! I feel bad leaving this in an incomplete state for so long. However, I will not be able to return to the project for right now, maybe in the future.

At this point it's probably better someone else finishes resolving this pull request considering Im a year out of touch (seeing as the stale flag was removed, I think that's what's happening anyways)

**Comment by janisozaur** [comment]
Yeah, we figured. No worries, thanks for the PR anyway.

**Comment by janisozaur** [comment]
I've done all I wanted for this PR. I don't plan to address any other comments myself.

@duncanspumpkin can you take this PR over finish line?

File: test/tests/Pathfinding.cpp [hunk 8]
```diff
@@ -86,7 +87,7 @@ class PathfindingTestBase : public testing::Test
 
         // Pick the direction the peep should initially move in, given the goal position.
         // This will also store the goal position and initialize pathfinding data for the peep.
-        gPeepPathFindGoalPosition = goal;
+        peep->SetPathfindGoal(TileCoordsXYZ(goal.ToCoordsXYZ()));
```
Comment (ZehMatt) [inline test/tests/Pathfinding.cpp:90]: goal is already TileCoordsXYZ, not sure whats going on here.

File: src/openrct2/peep/GuestPathfinding.cpp [hunk 9]
```diff
@@ -854,17 +850,15 @@ static void PeepPathfindHeuristicSearch(
                 }
                 else
                 { // numEdges == 2
-                    if (tileElement->AsPath()->IsQueue()
-                        && tileElement->AsPath()->GetRideIndex() != gPeepPathFindQueueRideIndex)
+
+                    auto guest = peep.As<Guest>();
```
Comment (ZehMatt) [inline src/openrct2/peep/GuestPathfinding.cpp:854]:
```suggestion
                    auto* guest = peep.As<Guest>();
```

**Review by ZehMatt — CHANGES_REQUESTED** [review]
Left a few comments, this one needs a final sweep but looks mostly okay, haven't tested this yet.

**Comment by ZehMatt** [comment]
@LordOfLunacy are you still working on this?

**Comment by ZehMatt** [comment]
> [#17232 (comment)](https://github.com/OpenRCT2/OpenRCT2/pull/17232#issuecomment-1531517174)
> 
> No unfortunately.

Alright thanks, and no worries.

**Pull request closed without merging**